lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -5296,16 +5296,35 @@ FIRCircuitParser::parseModuleBody(DeferredModuleToParse &deferredModule) {
 
   // Convert print-encoded verifications after parsing.
 
-  // It is dangerous to modify IR in the walk, so accumulate printFOp to
-  // buffer.
-  SmallVector<PrintFOp> buffer;
-  deferredModule.moduleOp.walk(
-      [&buffer](PrintFOp printFOp) { buffer.push_back(printFOp); });
-
-  for (auto printFOp : buffer) {
-    auto result = circt::firrtl::foldWhenEncodedVerifOp(printFOp);
-    if (failed(result))
-      return result;
+  {
+    // Gather PrintFOps and then process to avoid walking while mutating.
+    SmallVector<PrintFOp> buffer;
+    deferredModule.moduleOp.walk(
+        [&buffer](PrintFOp printFOp) { buffer.push_back(printFOp); });
+
+    size_t numVerifPrintfs = 0;
+    std::optional<Location> printfLoc;
+    for (auto printFOp : buffer) {
+      auto loc = printFOp.getLoc();
+      auto result = circt::firrtl::foldWhenEncodedVerifOp(printFOp);
+      if (failed(result))
+        return failure();
+      if (*result) {
+        ++numVerifPrintfs;
+        if (!printfLoc)
+          printfLoc = loc;
+      }
+    }
+    if (numVerifPrintfs > 0) {
+      auto diag =
+          mlir::emitWarning(deferredModule.moduleOp.getLoc(),
+                            "module contains ")
+          << numVerifPrintfs
```

Comment:
Can we open an issue for deprecation of printf-encoded verification and mention the issue in the warning message so that users can know what to do for migrations?

Reply:
Put up an issue here: https://github.com/llvm/circt/issues/6970 .

Reply:
Link added to diagnostic.